policy/handler.go

```diff
@@ -28,160 +68,283 @@ type Handler struct {
+			Policy:         h.policy,
+		}, check)
+
+		h.checkRunners = append(h.checkRunners, runner)
 	}
 
+	return nil
+}
+
+//	Run starts the handler for the given policy
+//
+// This function blocks until the context provided is canceled.
+func (h *Handler) Run(ctx context.Context) {
 	h.ticker = time.NewTicker(h.policy.EvaluationInterval)
 	defer h.ticker.Stop()
 
 	h.log.Trace("starting policy handler")
+
 	for {
 		select {
 		case <-ctx.Done():
 			h.log.Error("stopping policy handler due to context done")
+			h.errChn <- ctx.Err()
 			return
 
 		case updatedPolicy := <-h.updatesCh:
-			err := updatedPolicy.Validate()
-			if err != nil {
-				h.errChn <- fmt.Errorf("invalid policy: %v", err)
+			// The policy can be nil if the channel is closed meaning the
+			// handler is being removed, let the context cancelation take
+			// care of returning.
+			if updatedPolicy == nil {
+				continue
 			}
 
 			h.applyMutators(updatedPolicy)
 			h.updateHandler(updatedPolicy)
 
 		case <-h.ticker.C:
-			eval, err := h.handleTick(ctx, h.policy)
+
+			currentStatus, err := h.runTargetStatus()
 			if err != nil {
-				h.errChn <- fmt.Errorf("handler: unable to process policy %v", err)
+				h.errChn <- fmt.Errorf("handler: failed to get target status for target: %s, %w",
+					h.policy.Target.Name, err)
+			}
+
+			// A nil status indicates the target doesn't exist, log and return.
+			if currentStatus == nil {
+				h.errChn <- errTargetNotFound
+				continue
 			}
 
-			if eval != nil {
-				config.EvalsChannel <- eval
+			if currentStatus != nil && !currentStatus.Ready {
+				h.log.Debug("skipping evaluation, target not ready")
+				continue
 			}
 
-		case ts := <-h.cooldownCh:
-			// Enforce the cooldown which will block until complete.
-			if !h.enforceCooldown(ctx, ts) {
-				// Context was canceled, return to stop the handler.
-				return
+			currentCount := currentStatus.Count
+			action, err := h.CalculateNewCount(ctx, currentCount)
+			if err != nil {
+				h.errChn <- fmt.Errorf("handler: unable to execute policy ID: %s, %w",
+					h.policy.ID, err)
+				continue
+			}
+
+			// Canonicalize action so plugins don't have to.
+			action.Canonicalize()
+			action.CapCount(h.policy.Min, h.policy.Max)
+
+			h.log.Info("calculating scaling target", "from", currentCount, "to",
+				action.Count, "reason", action.Reason, "meta", action.Meta)
+
+			if !scalingNeeded(action, currentCount) {
+				h.log.Info("skipping scaling, no action needed")
+				continue
+			}
+
+			h.UpdateNextAction(action)
+
+			if nowFunc().After(h.OutOfCooldownOn()) {
+				h.UpdateState(StateIdle)
+			}
+
+			switch h.State() {
+			case StateCooldown:
+				h.log.Info("skipping scaling, policy still on cooldown", "remaining",
+					time.Until(h.OutOfCooldownOn()))
+
+			case StateWaitingTurn:
+				h.log.Debug("updating action, waiting to execute")
+
+			case StateScaling:
+				h.log.Info("skipping scaling, target still scaling")
+
+			case StateIdle:
+				go func() {
```

Comment:
We might want to think about moving the `h.UpdateState(StateWaitingTurn)` call out of the `WaitAndScale` method and before starting this goroutine. Consider the following scenario:

* ticker fires
* this `select` receives on `<-h.ticker.C` 
* `runTargetStatus` takes a long time, enough that the evaluation period passes
* ticker fires
* we start the `WaitAndScale` goroutine
* we go back to the top of the loop
* this `select` receives on `<-h.ticker.C` 
* we start another `WaitAndScale` goroutine
* the first `WaitAndScale` gets scheduled and grabs the lock

This is incredibly unlikely because the second `runTargetStatus` would have to return almost immediately to hit this race, but it's much like calling `sync.WaitGroup.Add(1)` before starting a goroutine.

Reply:
The change you've made doesn't address this because the `h.updateState(StateWaitingTurn)` is still happening inside the goroutine.